d2graph/d2graph.go

```diff
@@ -898,17 +898,36 @@ func appendTextDedup(texts []*d2target.MText, t *d2target.MText) []*d2target.MTe
 func (g *Graph) SetDimensions(mtexts []*d2target.MText, ruler *textmeasure.Ruler, fontFamily *d2fonts.FontFamily) error {
 	for _, obj := range g.Objects {
 		obj.Box = &geo.Box{}
-		// TODO fix edge cases for unnamed class etc
-		// Image shapes can set their own widths/heights
-		if obj.Attributes.Label.Value == "" && obj.Attributes.Shape.Value != d2target.ShapeImage {
-			obj.Width = 100
-			obj.Height = 100
-			continue
+
+		var setWidth int
+		var setHeight int
+		if obj.Attributes.Width != nil {
+			setWidth, _ = strconv.Atoi(obj.Attributes.Width.Value)
+		}
+		if obj.Attributes.Height != nil {
+			setHeight, _ = strconv.Atoi(obj.Attributes.Height.Value)
+		}
+		shapeType := strings.ToLower(obj.Attributes.Shape.Value)
+
+		switch shapeType {
+		case d2target.ShapeClass,
+			d2target.ShapeSQLTable,
+			d2target.ShapeCode,
+			d2target.ShapeImage,
+			d2target.ShapeText:
+			// edge cases for unnamed class, etc
+		default:
+			if obj.Attributes.Label.Value == "" && setWidth == 0 && setHeight == 0 {
```

Comment:
these should be split up. what if i only set width or only set height? can u add a test for that

Reply:
added a test and refactored so this is no longer needed